lib/Conversion/FIRRTLToLLHD/FIRRTLToLLHD.cpp

```diff
@@ -0,0 +1,270 @@
+    // of a port's aggregate type.
+    bool is_flip = false;
+    firrtl::FIRRTLType type = port.second;
+    if (auto flip_type = port.second.dyn_cast<firrtl::FlipType>()) {
+      is_flip = true;
+      type = flip_type.getElementType();
+    }
+
+    // Convert the type. We keep things simple for the time being.
+    auto width = type.getBitWidthOrSentinel();
+    if (width < 0) {
+      module.emitError() << "port " << port.first << " has unsupported type "
+                         << port.second;
+      signalPassFailure();
+      continue;
+    }
+    auto conv_type = SigType::get(builder->getIntegerType(width));
+
+    // Add to the list of inputs or outputs, depending on flip state.
+    if (is_flip) {
+      outs.push_back(conv_type);
+      out_names.push_back(port.first);
+      out_indices.push_back(i);
+    } else {
+      ins.push_back(conv_type);
+      in_names.push_back(port.first);
+      in_indices.push_back(i);
+    }
+  }
+
+  // Concatenate inputs and outputs and mark the split point for the entity.
+  // Then assemble the entity signature type.
+  auto num_ins = ins.size();
+  ins.append(outs.begin(), outs.end());
+  in_names.append(out_names.begin(), out_names.end());
+  in_indices.append(out_indices.begin(), out_indices.end());
+  auto entity_type = builder->getFunctionType(ins, llvm::None);
+
+  // Create an LLHD entity for this module.
+  auto entity = builder->create<EntityOp>(module.getLoc(), num_ins);
+  entity.setName(module.getName());
+  entity.setAttr("type", TypeAttr::get(entity_type));
+  EntityOp::ensureTerminator(entity.body(), *builder, entity.getLoc());
+
+  // Populate the arguments for the entity. This includes initializing the value
+  // mapping table with correspondences of the FIRRTL module arguments to the
+  // LLHD entity inputs/outputs. These are likely to be in a different order,
+  // such that we use the indirection indices gathered above.
+  auto args = entity.body().addArguments(ins);
+  for (auto arg : llvm::zip(args, in_indices)) {
+    auto firrtl_arg = module.getBodyBlock()->getArgument(std::get<1>(arg));
+    auto llhd_arg = std::get<0>(arg);
+    LLVM_DEBUG(llvm::dbgs()
+               << "Map FIRRTL port " << firrtl_arg.getArgNumber()
+               << " to LLHD port " << llhd_arg.getArgNumber() << "\n");
+    valueMapping[firrtl_arg] = llhd_arg;
+  }
+
+  // Populate the entity.
+  builder->setInsertionPoint(entity.body().front().getTerminator());
+  for (auto &op : module.getBodyBlock()->getOperations()) {
+    if (failed(dispatchVisitor(&op))) signalPassFailure();
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Conversion Helpers
+//===----------------------------------------------------------------------===//
+
+/// Get the LLHD value that has been previously emitted for a FIRRTL operand.
+Value FIRRTLToLLHDPass::getConvertedValue(Value operand) {
+  auto it = valueMapping.find(operand);
+  if (it == valueMapping.end()) {
```

Comment:
Seems like this should be an assert, since it is a local guarantee of this code?

Reply:
Yeah excellent point; fix coming up.